### PR TITLE
CAS-486 Bulk call to delius api to get offender details when get all applications

### PR DESCRIPTION
**Pull request by muhammad-elabdulla**
This [PR CAS-486](https://dsdmoj.atlassian.net/browse/CAS-486) currently when get all application is calling delius api endpoint for each crn. This PR is to change the call to delius api to bulk of crns

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt [hunk 1]
```diff
@@ -671,22 +671,6 @@ class OffenderServiceTest {
...
-      status = HttpStatus.FORBIDDEN,
-      method = HttpMethod.GET,
-      path = "/secure/offenders/crn/$crn/user/$deliusUsername/userAccess",
-      body = "{\"userRestricted\":false,\"userExcluded\":true,\"exclusionMessage\":\"You are excluded\"}",
-    )
-
-    val result = offenderService.canAccessOffender(deliusUsername, crn)
-
-    assertThat(result).isFalse()
-  }
```
Comment (davidatkinsuk) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt:688]: should we not have unit tests for the new function?

**Review by davidatkinsuk — APPROVED** [review]
Functionally looks good, with a few proposals in comments

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt (hunk 1 above)
Comment (muhammad-elabdulla) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt:688]: This is for the old function when it return ClientResult.Failure.StatusCode and HttpStatus.FORBIDDEN is not needed when using delius

Comment (davidatkinsuk) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt:688]: I agree this needs removing but we have a new function that doesn't appear to have unit tests? 

``` fun canAccessOffenders(username: String, crns: List<String>): Map<String, Boolean> ```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/ApplicationService.kt [hunk 2]
```diff
@@ -123,10 +123,11 @@ class ApplicationService(
...
 
-    return applicationSummaries
-      .filter {
-        offenderService.canAccessOffender(userDistinguishedName, it.getCrn())
-      }
+    val crns = applicationSummaries.map { it.getCrn() }.distinct()
+    val offendersAccess = offenderService.canAccessOffenders(userDistinguishedName, crns)
+    return applicationSummaries.filter {
+      offendersAccess.containsKey(it.getCrn()) && offendersAccess[it.getCrn()] == true
+    }
```
[the remark this answers is not in the document]
Comment (muhammad-elabdulla) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/ApplicationService.kt:130]: I changed the integration tests

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/OffenderService.kt [hunk 3]
```diff
@@ -357,29 +357,32 @@ class OffenderService(
...
+          valueTransform = { it },
+        )
+
+        crns.associateBy(
+          keySelector = { it },
+          valueTransform = { crn ->
+            val access = crnToAccessResult[crn]
+
+            if (access != null) {
+              !access.userExcluded && !access.userRestricted
```
[the remark this answers is not in the document]
Comment (muhammad-elabdulla) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/OffenderService.kt]: done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt (hunk 1 above)
Comment (muhammad-elabdulla) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/OffenderServiceTest.kt:688]: add unitests

**Pull request merged**